app/models/service_template.rb

```diff
@@ -33,6 +33,25 @@ class ServiceTemplate < ApplicationRecord
   virtual_has_one :custom_actions, :class_name => "Hash"
   virtual_has_one :custom_action_buttons, :class_name => "Array"
 
+  # almost feels like service_templates
+  def children
+    klass = self.class
+    service_resources.flat_map do |s|
+      s.resource.kind_of?(klass) ? s.resource : []
+    end
+  end
+
+  def descendants
+    klass = self.class
+    service_resources.flat_map do |s|
```

Comment:
maybe try:

```ruby
service_resources.where(:resource_type => base_class.name).map { |s| s.resource }.flat_map { |r| r.descendants + [r] }
```

this will not use cache. scope instead? Not sure how many times traversed. (think `service_resources` are accessed again soon. so best to just load them - verify)